applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h

```diff
@@ -0,0 +1,435 @@
+        double E2 = GetE2(T);
+        double* Emin = reinterpret_cast<double*>(&E1);
+        double* Emax = reinterpret_cast<double*>(&E2);
+        int M0 = static_cast<int>(subspace_size);
+        double* E = reinterpret_cast<double*>(tmp_eigenvalues.data().begin());
+        double* X = reinterpret_cast<double*>(tmp_eigenvectors.data().begin());
+        int M;
+        double* res = reinterpret_cast<double*>(residual.data().begin());
+        int info;
+
+        // call feast
+        auto feast = CallFeast(Ti, TSymmetric);
+        feast(&UPLO, &N, A, IA.data(), JA.data(), B, IB.data(), JB.data(), fpm, &epsout, &loop, Emin, Emax, &M0, E, X, &M, res, &info);
+
+        KRATOS_ERROR_IF(info < 0 || info > 99) << "FEAST encounterd error " << info << ". Please check FEAST output." << std::endl;
+        KRATOS_INFO_IF("FeastEigensystemSolver", info > 1 && info < 6) << "FEAST finished with warning " << info << ". Please check FEAST output." << std::endl;
+        KRATOS_ERROR_IF(info == 1) << "FEAST finished with warning " << info << ", no eigenvalues could be found in the given search interval." << std::endl;
+        KRATOS_ERROR_IF(info == 7) << "FEAST finished with warning " << info << ", no extremal eigenvalues could be found. Please check FEAST output." << std::endl;
+
+        // truncate non-converged results
+        tmp_eigenvalues.resize(M, true);
+        tmp_eigenvectors.resize(system_size, M, true);
+
+        RetrieveEigensolution<OutputVectorType, VectorType>(tmp_eigenvalues, rEigenvalues, tmp_eigenvectors, rEigenvectors);
+    }
+
+    DenseOutputMatrixType& GetEigenvectorSolution()
+    {
+        KRATOS_ERROR_IF( mpEigenvectorMatrix == nullptr ) << "The eigenvector matrix is not initialized. This means, that the real general generalized " <<
+            "eigenvalue problem did not yield complex results or the solution has not yet been performed" << std::endl;
+        return *mpEigenvectorMatrix;
+    }
+
+    OutputVectorType& GetEigenvalueSolution()
+    {
+        KRATOS_ERROR_IF( mpEigenvectorMatrix == nullptr ) << "The eigenvector matrix is not initialized. This means, that the real general generalized " <<
+            "eigenvalue problem did not yield complex results or the solution has not yet been performed" << std::endl;
+        return *mpEigenvalueVector;
+    }
+
+    /**
+     * Print information about this object.
+     */
+    void PrintInfo(std::ostream &rOStream) const override
+    {
+        rOStream << "FEASTEigensystemSolver";
+    }
+
+    /**
+     * Print object's data.
+     */
+    void PrintData(std::ostream &rOStream) const override
+    {
+    }
+
+  private:
+
+    typename TDenseSpaceTypeOut::MatrixPointerType mpEigenvectorMatrix;
+
+    typename TSparseSpaceTypeOut::VectorPointerType mpEigenvalueVector;
+
+    typedef void (feast_ptr)(char*, int*, double*, int*, int*, double*, int*, int*, int*, double*, int*, double*, double*, int*, double*, double*, int*, double*, int*);
```

Comment:
wow some explanation as to what is happening here would really help

Reply:
The FEAST functions for symmetric and unsymmetric eigenvalue problems do not have the same signature; for the symmetric case, the first parameter is a char, the rest are the same for the symmetric and general case.

Here I define a function pointer with the signature of the symmetric FEAST function (which is longer) and bind the functions providing all 19 arguments for the symmetric case (dfeast_scsrgv and zfeast_scsrgv) while only the last 18 arguments for the general case.

With these placeholders `_1, ..., _N` I can basically change the order of arguments I provided in the call of the function pointer and also completely omit some parameters. (I find the examples on https://en.cppreference.com/w/cpp/utility/functional/bind quite descriptive).

Should I add this as comment in the code?

Reply:
> Should I add this as comment in the code?

That would be good!

Reply:
ok done